### PR TITLE
feat(model): support postgres schema.

### LINKED ISSUE #2755: Indices trying to be created second time
- Gitea version (or commit ref): `1.2.1-linux-amd64`
- Git version:  `2.8.4`
- Operating system: `Centos 7 (7.4.1708)`
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist: [gitea.log](https://github.com/go-gitea/gitea/files/1404111/gitea.log)

## Description
**Database setup**: 
- Postgresql database (version 9.6) 
- database instance `appsdb`
- user `gitea`, created schema `gitea`
- `SEARCH_PATH` set to `'$user'` for `gitea` user (no public in `SEARCH_PATH`)  with command `ALTER ROLE gitea SET search_path TO "$user";`

Problem 1: After filling install form, tables were created, but instalation fails with log message
> [...itea/routers/init.go:56 GlobalInit()] [E] Failed to initialize ORM engine: sync database struct error: pq: > relation "UQE_user_lower_name" already exists

When I manually deleted all database tables, `gitea` started without problem (tables were recreated). I had to manually add user (`gitea admin create-user`) and then I was able to login.

Problem 2: After steps from section above, I've stopped `gitea`. The, attemt to start `gitea web` again failed with the same error as during instalation (`'... relation "UQE_user_lower_name" already exists...'`).
I've tried to drop index, but the same problem occured with another index. After several iterations I've dropped all indices (except primary keys); then `gitea web` started and recreated all deleted indices.

However, to drop all indices before each start is not usable.

### Possibly relevant info
I've found several links which seems to provide some relevant information:
- https://github.com/gogits/gogs/issues/3359
- https://github.com/go-xorm/xorm/issues/436
- https://github.com/go-xorm/xorm/issues/361
- https://github.com/go-xorm/xorm/issues/379

### PR DESCRIPTION
**Pull request by appleboy**
fix https://github.com/go-gitea/gitea/issues/2755

create new database and start the gitea.

```
e_base", "has_merged", "merged_commit_id", "merger_id", "merged_unix" FROM "pull_request" WHERE status = $1 []interface {}{1}
2018/01/11 15:10:55 [.../runtime/asm_amd64.s:2337 goexit()] [E] ArchiveClean: pq: relation "repository" does not exist
2018/01/11 15:10:55 [.../runtime/asm_amd64.s:2337 goexit()] [E] DeletedBranchesCleanup: pq: relation "deleted_branch" does not exist
2018/01/11 15:10:55 [I] [SQL] SELECT repo.id FROM "repository" repo WHERE repo.num_stars!=(SELECT COUNT(*) FROM "star" WHERE repo_id=repo.id)
2018/01/11 15:10:55 [I] [SQL] SELECT label.id FROM "label" WHERE label.num_issues!=(SELECT COUNT(*) FROM "issue_label" WHERE label_id=label.id)
2018/01/11 15:10:55 [...gitea/models/repo.go:2259 CheckRepoStats()] [E] Select repository count 'num_stars': pq: relation "repository" does not exist
2018/01/11 15:10:55 [.../runtime/asm_amd64.s:2337 goexit()] [E] DeliverHooks: pq: relation "hook_task" does not exist
2018/01/11 15:10:55 [I] [SQL] SELECT "user".id FROM "user" WHERE "user".num_repos!=(SELECT COUNT(*) FROM "repository" WHERE owner_id="user".id)
2018/01/11 15:10:55 [...gitea/models/repo.go:2259 CheckRepoStats()] [E] Select label count 'num_issues': pq: relation "label" does not exist
2018/01/11 15:10:55 [...gitea/models/repo.go:2259 CheckRepoStats()] [E] Select user count 'num_repos': pq: relation "user" does not exist
2018/01/11 15:10:55 [I] [SQL] SELECT "issue".id FROM "issue" WHERE "issue".num_comments!=(SELECT COUNT(*) FROM "comment" WHERE issue_id="issue".id AND type=0)
2018/01/11 15:10:55 [I] [SQL] SELECT repo.id FROM "repository" repo WHERE repo.num_closed_issues!=(SELECT COUNT(*) FROM "issue" WHERE repo_id=repo.id AND is_closed=$1AND is_pull=$2) []interface {}{true, false}
2018/01/11 15:10:55 [...gitea/models/repo.go:2259 CheckRepoStats()] [E] Select issue count 'num_comments': pq: relation "issue" does not exist
2018/01/11 15:10:55 [.../runtime/asm_amd64.s:2337 goexit()] [E] Select repository count 'num_closed_issues': pq: relation "repository" does not exist
2018/01/11 15:10:55 [I] [SQL] SELECT repo.id FROM "repository" repo WHERE repo.num_forks!=(SELECT COUNT(*) FROM "repository" WHERE fork_id=repo.id)
2018/01/11 15:10:55 [.../runtime/asm_amd64.s:2337 goexit()] [E] Select repository count 'num_forks': pq: relation "repository" does not exist
2018/01/11 15:10:55 [...gitea/models/pull.go:1161 TestPullRequests()] [E] Find Checking PRs%!(EXTRA *pq.Error=pq: relation "pull_request" does not exist)
2018/01/11 15:10:55 Serving [::]:3000 with pid 24634
2018/01/11 15:10:55 [I] Listen: http://0.0.0.0:3000
```

need @lunny help but it is working for `public` schema.

### CODE REVIEW

**Comment by appleboy** [comment]
update status:

Try the latest version of Gitea source code with update the latest `xorm` version and get the following error message:

```
WHERE c.relkind = 'r'::char AND c.relname = $1 AND s.table_schema = $2 AND f.attnum > 0 ORDER BY f.attnum; [oauth2_session public]
2018/05/01 09:04:02 [I] [SQL] SELECT indexname, indexdef FROM pg_indexes WHERE tablename=$1 AND schemaname=$2 [oauth2_session public]
2018/05/01 09:04:02 [I] [SQL] CREATE TABLE IF NOT EXISTS "oauth2_session" ("id" VARCHAR(400) PRIMARY KEY NOT NULL, "data" TEXT NULL, "created_unix" BIGINT NULL, "updated_unix" BIGINT NULL, "expires_unix" BIGINT NULL)
2018/05/01 09:04:02 [I] [SQL] CREATE INDEX "IDX_oauth2_session_expires_unix" ON "oauth2_session" ("expires_unix")
2018/05/01 09:04:02 [...itea/routers/init.go:64 GlobalInit()] [E] Failed to initialize OAuth2 support: pq: relation "IDX_oauth2_session_expires_unix" already exists
```

screen shot 2018-04-30 at 8 58 29 am

cc @lunny

**Comment by appleboy** [comment]
build errors:

1. https://drone.gitea.io/go-gitea/gitea/316/8
2. https://drone.gitea.io/go-gitea/gitea/316/10 (integration testing with postgres)

**Comment by appleboy** [comment]
@lunny Done. This PR can be review now.

File: models/models.go [hunk 1]
```diff
@@ -352,3 +383,11 @@ func DumpDatabase(filePath string, dbType string) error {
 	}
 	return x.DumpTablesToFile(tbs, filePath)
 }
+
+func schenaQuery(sql string) string {
```
Comment (techknowlogick) [inline models/models.go]: Is `schena` a typo? It is used elsewhere so it might not be, but I feel as though this should have an `m` not an `n`.

Comment (appleboy) [reply models/models.go]: this is a typo.

Comment (appleboy) [reply models/models.go]: See: https://github.com/go-gitea/gitea/pull/3348/commits/047e8a258193f914ba9b66bd9bd1d80a127aefd3

**Comment by lunny** [comment]
But CI failed. @appleboy it seems related.

**Comment by lunny** [comment]
see https://cdn.discordapp.com/attachments/323702390760931339/441956787625787393/unknown.png

**Comment by techknowlogick** [comment]
@appleboy please resolve conflict

**Comment by lkraider** [comment]
Can this be reviewed for merge? Using schemas makes the management of multiple DBs easier.

**Comment by techknowlogick** [comment]
@lkraider conflicts need to be resolved first

**Comment by lunny** [comment]
Please resolve the conflicts.

**Comment by sapk** [comment]
It seems that this PR will be superseeded by #8819

**Pull request closed without merging**